vector/src/main/java/im/vector/adapters/VectorMessagesAdapter.java

```diff
@@ -1969,4 +2160,163 @@ public boolean onMenuItemClick(final MenuItem item) {
+        MessageRow eventGroupRow = null;
+
+        if (supportMessageRowMerge(row)) {
+            if ((getCount() > 0) && (getItem(0).getEvent() instanceof EventGroup) && ((EventGroup) getItem(0).getEvent()).canAddRow(row)) {
+                eventGroupRow = getItem(0);
+            }
+
+            if (null == eventGroupRow) {
+                eventGroupRow = new MessageRow(new EventGroup(mContext.getApplicationContext(), mHiddenEventIds), null);
+                mEventGroups.add((EventGroup) eventGroupRow.getEvent());
+                super.insert(eventGroupRow, 0);
+                mEventRowMap.put(eventGroupRow.getEvent().eventId, row);
+            }
+
+            ((EventGroup) eventGroupRow.getEvent()).addToFront(row);
+            updateHighlightedEventId();
+        }
+
+        return (null != eventGroupRow);
+    }
+
+    /**
+     * Add a MessageRow into an EventGroup (if it is possible)
+     *
+     * @param row the row to added
+     */
+    private void addToEventGroup(MessageRow row) {
+        if (supportMessageRowMerge(row)) {
+            MessageRow eventGroupRow = null;
+
+            // search backward the EventGroup event
+            for (int i = getCount() - 1; i >= 0; i--) {
+                MessageRow curRow = getItem(i);
+
+                if (curRow.getEvent() instanceof EventGroup) {
+                    // the event can be added (same day ?)
+                    if (((EventGroup) curRow.getEvent()).canAddRow(row)) {
+                        eventGroupRow = curRow;
+                    }
+                    break;
+                } else
+                    // there is no more room member events
+                    if (!TextUtils.equals(curRow.getEvent().getType(), Event.EVENT_TYPE_STATE_ROOM_MEMBER)) {
+                        break;
+                    }
+            }
+
+            if (null == eventGroupRow) {
+                eventGroupRow = new MessageRow(new EventGroup(mContext.getApplicationContext(), mHiddenEventIds), null);
+                super.add(eventGroupRow);
+                mEventGroups.add((EventGroup) eventGroupRow.getEvent());
+                mEventRowMap.put(eventGroupRow.getEvent().eventId, eventGroupRow);
+            }
+
+            ((EventGroup) eventGroupRow.getEvent()).add(row);
+            updateHighlightedEventId();
+        }
+    }
+
+    /**
+     * Remove a message row from the known event groups
+     *
+     * @param row the message row
+     * @return true if the message has been removed
+     */
+    private void removeFromEventGroup(MessageRow row) {
+        if (supportMessageRowMerge(row)) {
+            String eventId = row.getEvent().eventId;
+            for (EventGroup eventGroup : mEventGroups) {
+                if (eventGroup.contains(eventId)) {
+                    eventGroup.removeByEventId(eventId);
+
+                    if (eventGroup.isEmpty()) {
+                        mEventGroups.remove(eventGroup);
```

Comment:
Removing the object from the list while iterating on it can cause issues. I think an iterator would be better.
https://stackoverflow.com/questions/1196586/calling-remove-in-foreach-loop-in-java